imports/plugins/core/product/server/no-meteor/mutations/cloneProductVariants.js

```diff
@@ -0,0 +1,125 @@
+ * @return {Array} list with cloned variant Ids
+ */
+export default async function cloneProductVariants(context, input) {
+  inputSchema.validate(input);
+  const { collections, userHasPermission } = context;
+  const { Products } = collections;
+  const { shopId, variantIds } = input;
+
+  if (!userHasPermission(["createProduct", "product/admin", "product/clone"], shopId)) {
+    throw new ReactionError("access-denied", "Access Denied");
+  }
+
+  // Check to make sure all variants are on the same shop
+  const count = await Products.find({ _id: { $in: variantIds }, type: "variant", shopId }).count();
+  if (count !== variantIds.length) throw new ReactionError("not-found", "One or more variants do not exist");
+
+  const newVariantIds = await Promise.all(variantIds.map(async (variantId) => {
+    const existingVariant = await Products.findOne({ _id: variantId });
+    const productId = existingVariant.ancestors[0];
+
+    const existingVariants = await Products.find({
+      $or: [
+        {
+          _id: variantId
+        },
+        {
+          ancestors: {
+            $in: [variantId]
+          },
+          isDeleted: false
+        }
+      ],
+      type: "variant"
+    }).toArray();
+
+    // If there are no variants, return empty array
+    if (existingVariants.length === 0) return null;
+
+    const variantNewId = Random.id();
+
+    // make sure that top level variant will be cloned first
+    const sortedVariants = _.sortBy(existingVariants, (sortedVariant) => sortedVariant.ancestors.length);
+
+    await Promise.all(sortedVariants.map(async (sortedVariant) => {
+      const originalVariantId = sortedVariant._id;
+      let type = "child";
+      const clonedVariantObject = {};
+      if (variantId === sortedVariant._id) {
+        type = "parent";
+        Object.assign(clonedVariantObject, sortedVariant, {
+          _id: variantNewId,
+          title: `${sortedVariant.title} - copy`,
+          optionTitle: `${sortedVariant.optionTitle} - copy`
+        });
+      } else {
+        const parentIndex = sortedVariant.ancestors.indexOf(variantId);
+        const ancestorsClone = sortedVariant.ancestors.slice(0);
+        // if variantId exists in ancestors, we override it by new _id
+        if (parentIndex >= 0) ancestorsClone.splice(parentIndex, 1, variantNewId);
+        Object.assign(clonedVariantObject, existingVariant, {
+          _id: Random.id(),
+          ancestors: ancestorsClone,
+          title: `${sortedVariant.title}`,
+          optionTitle: `${sortedVariant.optionTitle}`,
+          height: `${sortedVariant.height}`,
+          width: `${sortedVariant.width}`,
+          weight: `${sortedVariant.weight}`,
+          length: `${sortedVariant.length}`
+        });
+      }
+      delete clonedVariantObject.updatedAt;
+      delete clonedVariantObject.createdAt;
+
+      // Apply custom transformations from plugins.
+      for (const customFunc of context.getFunctionsOfType("mutateNewVariantBeforeCreate")) {
```

Comment:
I copied the `clone` functions as close as I could from the Meteor methods. I noticed that this `getFunctionsOfType` loop was in the `variant` clone method, but not in the `product` clone method, even the `variant` section of that method. Does it need to be added there?

Reply:
That might be an oversight. It seems like it should be in both, or possibly in neither since they are cloned from variants that were already mutated. It depends on what those functions do. Possibly there should be separate `mutateNewVariantBeforeCreate` and `mutateNewVariantAfterClone` for the most specificity.